src/Service/ProducerEventProcessor.cpp

```diff
@@ -0,0 +1,371 @@
+// Copyright (c) 2021 The Orbit Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#include "ProducerEventProcessor.h"
+
+#include <absl/container/flat_hash_map.h>
+
+#include "OrbitBase/Logging.h"
+#include "capture.pb.h"
+
+namespace orbit_service {
+
+namespace {
+
+using orbit_grpc_protos::AddressInfo;
+using orbit_grpc_protos::Callstack;
+using orbit_grpc_protos::CallstackSample;
+using orbit_grpc_protos::ClientCaptureEvent;
+using orbit_grpc_protos::FullAddressInfo;
+using orbit_grpc_protos::FullCallstackSample;
+using orbit_grpc_protos::FullGpuJob;
+using orbit_grpc_protos::FullTracepointEvent;
+using orbit_grpc_protos::FunctionCall;
+using orbit_grpc_protos::GpuDebugMarker;
+using orbit_grpc_protos::GpuJob;
+using orbit_grpc_protos::GpuQueueSubmission;
+using orbit_grpc_protos::InternedCallstack;
+using orbit_grpc_protos::InternedString;
+using orbit_grpc_protos::InternedTracepointInfo;
+using orbit_grpc_protos::IntrospectionScope;
+using orbit_grpc_protos::ModuleUpdateEvent;
+using orbit_grpc_protos::ProducerCaptureEvent;
+using orbit_grpc_protos::SchedulingSlice;
+using orbit_grpc_protos::ThreadName;
+using orbit_grpc_protos::ThreadStateSlice;
+using orbit_grpc_protos::TracepointEvent;
+
+template <typename T>
+class InternedPool final {
+ public:
+  InternedPool() = default;
+
+  // Return pair of <id, assigned>, where assigned is true if the entry was assigned a new id
+  // and false if returning id for already existing entry.
+  std::pair<uint64_t, bool> GetOrAssignId(const T& entry) {
+    absl::MutexLock lock(&entry_to_id_mutex_);
+    auto it = entry_to_id_.find(entry);
+    if (it != entry_to_id_.end()) {
+      return std::make_pair(it->second, false);
+    }
+
+    uint64_t new_id = id_counter_++;
+    auto [unused_it, inserted] = entry_to_id_.insert_or_assign(entry, new_id);
+    CHECK(inserted);
+    return std::make_pair(new_id, true);
+  }
+
+ private:
+  uint64_t id_counter_ = 1;  // 0 is reserved for invalid_id
```

Comment:
Do we need to define a constant for the invalid id?

Reply:
we have one for invalid function id, I guess we could do that for all ids, done.